shopify_function_macro/src/lib.rs

```diff
@@ -114,6 +115,158 @@ pub fn shopify_function(
+        "Shopify Functions must return a Result"
+    );
+    let PathArguments::AngleBracketed(generics) = &result.arguments else {
+        panic!("Shopify Function Result is missing generic arguments")
+    };
+    assert!(
+        generics.args.len() == 1,
+        "Shopify Function Result takes exactly one generic argument"
+    );
+    let GenericArgument::Type(ty) = generics.args.first().unwrap() else {
+        panic!("Shopify Function Result expects a type")
+    };
+    let Type::Path(path) = ty else {
+        panic!("Unexpected result type for Shopify Function Result")
+    };
+    &path.path.segments.last().as_ref().unwrap().ident
+}
+
+#[proc_macro_attribute]
+pub fn shopify_function_target(
+    attr: proc_macro::TokenStream,
+    item: proc_macro::TokenStream,
+) -> proc_macro::TokenStream {
+    let ast = parse_macro_input!(item as syn::ItemFn);
+    let args = parse_macro_input!(attr as ShopifyFunctionTargetArgs);
+
+    let name = &ast.sig.ident;
+    let export_ident = args.export.map_or(name.clone(), |export| Ident::new(export.value().as_str(), Span::mixed_site()));
+    let export_string = export_ident.to_string();
+
+    let query_path = args.query_path.expect("No value given for query_path");
+    let schema_path = args.schema_path.expect("No value given for schema_path");
+    let output_query_file_name = format!(".{}{}", &export_string, OUTPUT_QUERY_FILE_NAME);
+
+    let input_struct = generate_struct(
+        "Input",
+        query_path.value().as_str(),
+        schema_path.value().as_str(),
+    );
+    let output_struct = generate_struct(
+        "Output",
+        &output_query_file_name,
+        schema_path.value().as_str(),
+    );
+    let output_result_type = extract_shopify_function_return_type(&ast)
+        .to_token_stream()
+        .to_string();
+    let output_query = format!(
+        "mutation Output($result: {}!) {{\n    {}(result: $result)\n}}\n",
+        output_result_type, &export_string.to_case(Case::Camel)
+    );
+
+    write_output_query_file(&output_query_file_name, &output_query);
+
+    let input_stream = args
+        .input_stream
+        .map_or(quote! { std::io::stdin() }, |stream| {
+            stream.to_token_stream()
+        });
+    let output_stream = args
+        .output_stream
+        .map_or(quote! { std::io::stdout() }, |stream| {
+            stream.to_token_stream()
+        });
+
+    quote! {
+        pub mod #export_ident {
+            use super::*;
+            use std::io::Write;
+
+            #input_struct
+            #output_struct
+
+            #[shopify_function(
+                input_stream = #input_stream,
+                output_stream = #output_stream
+            )]
+            #ast
+
+            #[export_name = #export_string]
+            pub extern "C" fn export() {
+                main().unwrap();
+                #output_stream.flush().unwrap();
+            }
+        }
+        #[allow(dead_code)]
+        #ast
```

Comment:
Rather than emitting the same function twice, we can just import it from the generated module. You might have to adjust the `shopify_function` macro to inject the `pub` visibility to the function (i.e. overwriting `ast.vis`).

```suggestion
        use #export_ident::#name;
```